jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayAllMessagesActivity.java

```diff
@@ -0,0 +1,57 @@
+package com.e_conomic.jonfirstapp;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+import android.support.v7.app.AppCompatActivity;
+import android.util.Log;
+import android.widget.TextView;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.charset.Charset;
+
+
+public class DisplayAllMessagesActivity extends AppCompatActivity {
+
+    private String messageFileName;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_display_all_messages);
+
+        Intent intent = getIntent();
+        messageFileName = intent.getStringExtra(MainActivity.EXTRA_MESSAGE_FILENAME);
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        TextView textView = (TextView) findViewById(R.id.text_view_show_all_messages);
+        textView.setText(getAllMessages());
+    }
+
+
+    /** Retrieve all the previous messages written. */
+    private String getAllMessages() {
+        FileInputStream messageInputStream;
+        int tempByte;
+        StringBuffer allMessages = new StringBuffer("");
+
+        try {
+            messageInputStream = openFileInput(messageFileName);
+            while ((tempByte = messageInputStream.read()) != -1) {
```

Comment:
Yup it is very annoying :D They don't show a way to do it in the tutorial. There is a way to read multiple bytes at once using `FileInputStream.read(byte[] b)` (see [here](https://www.tutorialspoint.com/java/io/fileinputstream_read_byte.htm)), but I found that it sometimes (I think at the end of the file) it read some bytes twice! I couldn't figure out why, so I found two possibilities where the above is one of them. The other one is to check how many bytes are available and read accordingly. Both seem not so nice.. if you prefer the second option or if you have a way to do it nicer, please let me know :)